compiler/env/OMRAOTMethodHeader.cpp

```diff
@@ -0,0 +1,183 @@
+/*******************************************************************************
+ * Copyright (c) 2022, 2022 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at http://eclipse.org/legal/epl-2.0
+ * or the Apache License, Version 2.0 which accompanies this distribution
+ * and is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the
+ * Eclipse Public License, v. 2.0 are satisfied: GNU General Public License,
+ * version 2 with the GNU Classpath Exception [1] and GNU General Public
+ * License, version 2 with the OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+#include "env/AOTMethodHeader.hpp"
+#include "infra/Assert.hpp"
+#include "runtime/TRRelocationRecord.hpp"
+#include <stdlib.h>
+#include <string.h>
+
+TR::AOTMethodHeader*
+OMR::AOTMethodHeader::self()
+   {
+   return reinterpret_cast<TR::AOTMethodHeader*> (this);
+   }
+
+OMR::AOTMethodHeader::AOTMethodHeader(uint8_t* compiledCodeStart, uint32_t compiledCodeSize, uint8_t* relocationsStart, uint32_t relocationsSize):
+      _compiledCodeStart(compiledCodeStart),
+      _compiledCodeSize(compiledCodeSize),
+      _relocationsSize(relocationsSize)
+      {
+      self()->setRelocationsStart(relocationsStart);
+      };
+
+void
+OMR::AOTMethodHeader::serializeMethod(uint8_t* buffer,size_t bufferSize)
+   {
+   size_t allocSize = self()->sizeOfSerializedVersion();
+   TR_ASSERT(bufferSize >= allocSize, "Insufficient memory allocated for AOT Method serialization.\n");
+
+   uint8_t* ptr = buffer;
+
+   // Copy the size of the header to the buffer
+   memcpy(ptr,&allocSize,sizeof(size_t));
```

Comment:
have you considered declaring a "local" struct (with last field being a byte array of size 1) to make this code easier to write? i.e. you can just cast the pointer to a pointer to that struct and then write the fields (just like how the BinaryTempate classes work for relocation records).

Reply:
Hello Mark, thank you for correction suggestion, I have used the line of code you specified.  The serialization code is there mostly like an idea, I am inviting runtime developers to specify their own layout, without enforcing too much additional constructs. If you believe I should introduce OMRAOTMethodHeaderBinaryTemplate extensible class hierarchy, I can do that, but that looks like an overkill to me. Please let me know what you think is better  for OMR.

Reply:
I just meant to define a simple struct specific to your use case (rather than repeatedly bumping pointers manually by `sizeof(last thing you stored)`. I'm not suggesting to mandate the use of an extensible struct in downstream projects.